Cabal/tests/PackageTests/Tests.hs

```diff
@@ -0,0 +1,157 @@
+module PackageTests.Tests(tests) where
+
+import PackageTests.PackageTester
+
+import qualified PackageTests.BenchmarkStanza.Check
+import qualified PackageTests.TestStanza.Check
+import qualified PackageTests.DeterministicAr.Check
+import qualified PackageTests.TestSuiteTests.ExeV10.Check
+
+import Distribution.Simple.Utils (withFileContents)
+import Control.Monad
+
+import Data.Version
+import Test.Tasty (TestTree, testGroup, mkTimeout, localOption)
+import Test.Tasty.HUnit (testCase)
+
+-- TODO: turn this into a "test-defining writer monad".
+-- This will let us handle scoping gracefully.
+tests :: SuiteConfig -> [TestTree]
+tests config =
+  tail [ undefined
+
+  -- External tests
+  , tc "BenchmarkStanza"  PackageTests.BenchmarkStanza.Check.suite
+  , tc "TestStanza"       PackageTests.TestStanza.Check.suite
+  , tc "DeterministicAr"  PackageTests.DeterministicAr.Check.suite
+
+  -- Test suite tests
+  , testGroup "TestSuiteTests"
+    [ testGroup "ExeV10"
+      (PackageTests.TestSuiteTests.ExeV10.Check.tests config)
+    , testGroup "LibV09" $
+      let
+        tcs :: FilePath -> TestM a -> TestTree
+        tcs name m
+            = testCase name (runTestM config ("TestSuiteTests/LibV09")
+                                             (Just name) m)
+      in [ tcs "Build" $ cabal_build ["--enable-tests"]
+         , localOption (mkTimeout $ 10 ^ (8 :: Int))
+         . tcs "Deadlock" $ do
+            cabal_build ["--enable-tests"]
+            shouldFail $ cabal "test" []
+         ]
+    ]
+
+  -- Inline tests
+  , tc "BenchmarkExeV10" $ cabal_build ["--enable-benchmarks"]
+
+  , tc "BenchmarkOptions" $ do
```

Comment:
One minor complaint: It's a bit unclear what the tests are actually trying to test. E.g. is this test just trying to see if cabal accepts the options without failing badly? Is it trying to test whether benchmarks executables are actually produced? (etc.)

Clarifying comments would suffice, I think, but personally, I'm a big fan of doing a little more formal "setup/exercise/verify" stanzas. See for example the first bit of example code on http://xunitpatterns.com/Four%20Phase%20Test.html for the gist of the idea. Doing this slightly more formally with "setup", "exercise", "verify" combinators would be trivial. *Rationale:* I find that doing this formally (or at least semi-formally using always-identical comments) tends to make it a lot easier to figure out unfamiliar failing tests.